### DISCUSSION TITLE
The unnamed module with Qodana JVM (Java 17 and --add-exports)

### DISCUSSION THREAD

**Discussion by Cazayus** [JetBrains/Qodana#129] [Q&A] [answered]
Hello,

We are trying to use Qodana in our project and ran into some problems related to the JVM

1. The project is in Java 14 and by default Qodana JVM runs in Java 11 I think. 
The solution in #68 and #120 works but there is no JVM available for 14.
I believe that the option to choose which JVM/Langage level you want should be somewhere in the main documentation. 
The ability to choose a JVM 14 is perhaps not very important because it looks like most JVM between 8 and 17 are available and we can just pick a more recent one.

2. We then decided to try Qodana with Java 17 and it is reporting multiple correctness issue that I don't know how to solve. 
Basically, I need to pass multiple --add-exports arguments to compile our project and I don't know how to do the equivalent with Qodana JVM.
I tried to use the _JAVA_OPTIONS found in the documentation but I still get the same errors in the report about "Package 'XXX' is declared in module 'YYY', which does not export it to the unnamed module"
I can see the following message in the log saying : 

> Picked up _JAVA_OPTIONS: --add-exports=java.desktop/sun.awt=ALL-UNNAMED --add-exports=java.desktop/com.sun.java.swing.plaf.windows=ALL-UNNAMED --add-exports=java.desktop/sun.swing=ALL-UNNAMED --add-exports=java.desktop/com.sun.java.swing.plaf.motif=ALL-UNNAMED
WARNING: package com.sun.java.swing.plaf.windows not in java.desktop

Qodana sees something but does not use it like I would like it to.
I'm also surprised to see "package com.sun.java.swing.plaf.windows not in java.desktop", because it very much is as far as I can tell.

Is there any way to pass the --add-exports arguments ?

**Comment by Cazayus** [comment]
I think I know why Qodana reports "package com.sun.java.swing.plaf.windows not in java.desktop".
In the jdks mentionned in #68, it looks like there is different OpenJDK 17 flavor depending on your OS.
In the Windows one, I found the package, but not in the Linux one.

**Reply by brichbash** [reply]
You're right. Qodana images are based on Linux, thus they need JDKs compatible with it.

**Comment by brichbash** [comment, marked as answer]
Hello,

1. With the current approach you indeed can't download JDK 14, as it is not available on our CDN (you can't download it in IntelliJ IDEA as well). We're currently working on making JDK configuration easier in Qodana, and after it's implemented, we'll document it. Thank you for bringing this up.
2. Could you please try creating `.idea/compiler.xml` with the following content in your project root (you can also copy it from your machine if you have your project configured in IntelliJ IDEA):
```
<?xml version="1.0" encoding="UTF-8"?>
<project version="4">
  <component name="JavacSettings">
    <option name="ADDITIONAL_OPTIONS_STRING" value="--add-exports=java.desktop/sun.awt=ALL-UNNAMED --add-exports=java.desktop/com.sun.java.swing.plaf.windows=ALL-UNNAMED --add-exports=java.desktop/sun.swing=ALL-UNNAMED --add-exports=java.desktop/com.sun.java.swing.plaf.motif=ALL-UNNAMED" />
  </component>
</project>
```

**Reply by Cazayus** [reply]
Thank you, modifying `compiler.xml` worked !

I would say that this kind of tricky configuration should also be part of the "making JDK configuration easier" you are working on

**Reply by brichbash** [reply]
I have created [an issue](https://youtrack.jetbrains.com/issue/QD-4342/Add-an-option-to-pass-additional-parameters-to-JavaKotlin-compilers-via-qodanayaml) on our tracker regarding compiler configuration in Qodana. Feel free to vote for it to get notified when there are any updates.
